### PR TITLE
[MM-34204] Adding circle ci config, makefile, dockerfile and supporting scripts.

### PR DESCRIPTION
**Pull request by stylianosrigas**
#### Summary
Adding:
- Circle CI pipeline
- Makefile
- Genesis Dockerfile
- Supporting scripts

We can merge and start using those when https://github.com/mattermost/genesis/pull/2 gets merged.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34204

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+      environment:
+        POSTGRES_USER: genesis_test
+        POSTGRES_DB: genesis_test
+    executor:
+      name: default
+    steps:
+    - run:
+        name: Waiting for Postgres to be ready
+        command: |
+          for i in `seq 1 10`;
```
Comment (spirosoik) [inline .circleci/config.yml:76]: just for consistency should we make this a bash script?

Comment (spirosoik) [reply .circleci/config.yml:76]: or just use the [wait-for-it](https://github.com/vishnubob/wait-for-it)

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+MATTERMOST_GENESIS_IMAGE ?= mattermost/genesis:test
+MACHINE = $(shell uname -m)
+GOFLAGS ?= $(GOFLAGS:)
+BUILD_TIME := $(shell date -u +%Y%m%d.%H%M%S)
+BUILD_HASH := $(shell git rev-parse HEAD)
+
+################################################################################
+
+BUILD_HASH = $(shell git rev-parse HEAD)
+LDFLAGS += -X "github.com/mattermost/genesis/model.BuildHash=$(BUILD_HASH)"
```
Comment (spirosoik) [inline Makefile:27]: why not pass the commit hash from circleci env variables?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+OUTDATED_VER := master
+OUTDATED_BIN := go-mod-outdated
+OUTDATED_GEN := $(TOOLS_BIN_DIR)/$(OUTDATED_BIN)
+
+GOVERALLS_VER := master
+GOVERALLS_BIN := goveralls
+GOVERALLS_GEN := $(TOOLS_BIN_DIR)/$(GOVERALLS_BIN)
+
+GOLINT_VER := master
+GOLINT_BIN := golint
```
Comment (spirosoik) [inline Makefile:46]: for future reference, golanci-linter is a bit better and more comprehensive https://github.com/golangci/golangci-lint (we can add it as a ticket to investigate it)

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+## Runs lint against all packages.
+.PHONY: lint
+lint: $(GOLINT_GEN)
+	@echo Running lint
+	$(GOLINT_GEN) -set_exit_status ./...
+	@echo lint success
+
+## Runs govet against all packages.
+.PHONY: vet
+govet:
```
Comment (spirosoik) [inline Makefile:68]: very very nice

File: README.md [hunk 5]
```diff
@@ -1 +1,3 @@
-# genesis
\ No newline at end of file
+# Genesis
+
+This repository houses the open-source components of Mattermost Genesis. This tool is providing a service to provision AWS Accounts and manage VPC CIDR pools, VPC infrastructure and VPC peering functionality.
```
Comment (spirosoik) [inline README.md:3]: I can imagine that you have already in your list to document the details.

File: scripts/go_install.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+
+# Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+# See LICENSE.txt for license information.
+
+set -o errexit
```
Comment (spirosoik) [inline scripts/go_install.sh:6]: I often use xtrace mode to get more precise debug information

File: scripts/push-docker-pr.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+# Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+# See LICENSE.txt for license information.
+
+set -e
```
Comment (spirosoik) [inline scripts/push-docker-pr.sh:6]: I think for consistency will be good to use one format (short or long) for setting thes

File: scripts/push-docker.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+# Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+# See LICENSE.txt for license information.
+
+set -e
```
Comment (spirosoik) [inline scripts/push-docker.sh:6]: same here

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+  push-docker-pr:
+    executor:
+      name: default
+    steps:
+    - checkout
+    - setup_remote_docker
+    - run: make build-image
+    - run: ./scripts/push-docker-pr.sh
+    - run:
```
Comment (spirosoik) [inline .circleci/config.yml:96]: out of curiosity what made us to add the extra rerun?

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+      environment:
+        GIT_STRATEGY: none
+        CI_APPLICATION_REPOSITORY: mattermost/genesis
+        CLAIR_DB_CONNECTION_STRING: "postgresql://postgres:password@localhost:5432/postgres?sslmode=disable&statement_timeout=60000"
+        DOCKERFILE_PATH: "build/Dockerfile"
+    - image: arminc/clair-db:latest
+    steps:
+    - checkout
+    - run: |
+        export CI_APPLICATION_TAG="${CIRCLE_SHA1:0:7}"
```
Comment (spirosoik) [inline .circleci/config.yml:130]: I think you can do this as part of the run 
```
environment:
 CI_APPLICATION_TAG="${CIRCLE_SHA1:0:7}" <--- not sure if this works though, eg. in github actions doesn't if I recall
 DOCKER_USER=$DOCKER_USERNAME
```

**Review by spirosoik — COMMENTED** [review]
just minor a few comments, mostly curious actually. other than that looks good.

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    - checkout
+    - *restore_cache
+    - run: |
+        make unittest
+        make goverall
+    - *save_cache
+
+  test-postgres:
+    docker:
+    - image: circleci/golang:1.14.6
```
Comment (angeloskyratzakos) [inline .circleci/config.yml:63]: We use on purpose older version of golang?

File: .circleci/config.yml (hunk 1 above)
Comment (stylianosrigas) [reply .circleci/config.yml]: moved to its own script, thanks :)

File: Makefile (hunk 3 above)
Comment (stylianosrigas) [reply Makefile]: Nice, I like it. Moved to use golangci-lint

File: README.md (hunk 5 above)
Comment (stylianosrigas) [reply README.md:3]: Yeap the readme is added in another PR ;)

File: scripts/go_install.sh (hunk 6 above)
Comment (stylianosrigas) [reply scripts/go_install.sh:7]: All bash scripts should be consistent now :)

File: .circleci/config.yml (hunk 9 above)
Comment (stylianosrigas) [reply .circleci/config.yml]: circle ci had some issues before and was not able to push our images some times. I think its fixed now. Removed it ;)

File: .circleci/config.yml (hunk 11 above)
Comment (stylianosrigas) [reply .circleci/config.yml]: Nope mistake. Fixed ;)

File: .circleci/config.yml (hunk 10 above)
Comment (stylianosrigas) [reply .circleci/config.yml]: Yeap nice, changed it :)

**Review by spirosoik — APPROVED** [review]
very good job

**Comment by spirosoik** [comment]
let's resolve the conflicts only

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,242 @@
+version: 2.1
+executors:
+  default:
+    docker:
+    - image: circleci/golang:1.16.2
```
Comment (cpanato) [inline .circleci/config.yml:5]: can update to 1.16.3

File: .circleci/config.yml [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    - checkout
+    - *restore_cache
+    - run: |
+        make unittest
+        make goverall
+    - *save_cache
+
+  test-postgres:
+    docker:
+    - image: circleci/golang:1.16.2
```
Comment (cpanato) [inline .circleci/config.yml:63]: same here

File: Makefile [hunk 14]
```diff
@@ -0,0 +1,139 @@
+# Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+# See LICENSE.txt for license information.
+
+################################################################################
+##                             VERSION PARAMS                                 ##
+################################################################################
+
+## Docker Build Versions
+DOCKER_BUILD_IMAGE = golang:1.16.2
```
Comment (cpanato) [inline Makefile:9]: same

File: Makefile [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+DOCKER_BASE_IMAGE = alpine:3.12.4
```
Comment (cpanato) [inline Makefile:10]: can update to alpine:3.13.4

File: build/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+# See LICENSE.txt for license information.
+
+# Build the mattermost genesis
+ARG DOCKER_BUILD_IMAGE=golang:1.16.2
+ARG DOCKER_BASE_IMAGE=alpine:3.12.4
```
Comment (cpanato) [inline build/Dockerfile:6]: same here for both

**Review by cpanato — COMMENTED** [review]
nice work

File: scripts/push-docker.sh [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+else
+  echo "Pushing release $CIRCLE_TAG..."
+  TAG="$CIRCLE_TAG"
+fi
+
+echo $DOCKER_PASSWORD | docker login --username $DOCKER_USERNAME --password-stdin
+
+docker tag mattermost/mattermost-cloud:test mattermost/mattermost-cloud:$TAG
+
+docker push mattermost/mattermost-cloud:$TAG
```
Comment (angeloskyratzakos) [inline scripts/push-docker.sh:23]: this should tag `mattermost-cloud`? 🤔

Comment (stylianosrigas) [reply scripts/push-docker.sh]: Yeap nice catch, thanks :)

**Comment by stylianosrigas** [comment]
> nice work

thanks for the review buddy

**Review by angeloskyratzakos — APPROVED** [review]
LGTM

**Pull request merged**